src/microbe_stage/editor/EditorActionHistory.cs

```diff
@@ -191,6 +175,37 @@ public bool HexPlacedThisSession<THex>(THex hex)
         return History.OfType<HexPlacementActionData<THex>>().Any(a => a.PlacedHex == hex);
     }
 
+    /// <summary>
+    ///   Pick the data with which you can regain maximum MP to combine.
+    /// </summary>
+    /// <param name="currentData">The data you want to combine</param>
+    /// <param name="previousData">A list of data currentData may combine with</param>
+    /// <returns>
+    ///   Three-element tuple:
+    ///   CostDelta is non-positive, indicating the amount of MP you will regain;
+    ///   MinimumCostActionData is the action data to combine with (not yet combined);
+    ///   Mode is the interference mode currentData has with MinimumCostActionData.
+    /// </returns>
+    private static (int CostDelta, EditorCombinableActionData MinimumCostActionData, ActionInterferenceMode Mode)
+        GetLeastCostCombineTuple(EditorCombinableActionData currentData,
+            IEnumerable<EditorCombinableActionData> previousData)
+    {
+        return previousData.Select(data =>
+        {
+            var interferenceMode = currentData.GetInterferenceModeWith(data);
+
+            return (interferenceMode switch
+            {
+                ActionInterferenceMode.Combinable => ((EditorCombinableActionData)currentData.Combine(data))
```

Comment:
I'm a bit suspicious about the cast here. What guarantees this doesn't fail?

Reply:
What doesn't fail? The cast? EditorCom... combines EditorCom... always return a EditorCom..

Reply:
I read the code in my IDE, it's safe to call this as `currentData` so we can assume Combine returns a `EditorCombinableActionData`.